src/red_black_tree.jl

```diff
@@ -0,0 +1,327 @@
+# it has unique keys
+# leftChild has keys which are less than the node
+# rightChild has keys which are greater than the node
+# color is true if it's a Red Node, else it's false
+mutable struct RBTreeNode{K}
+    color::Bool 
+    data::Union{K, Nothing}
+    leftChild::Union{Nothing, RBTreeNode{K}}
+    rightChild::Union{Nothing, RBTreeNode{K}}
+    parent::Union{Nothing, RBTreeNode{K}}
+
+    RBTreeNode{K}() where K = new{K}(true, nothing, nothing, nothing, nothing)
+    
+    RBTreeNode{K}(d::K) where K = new{K}(true, d, nothing, nothing, nothing)
+end
+
+RBTreeNode() = RBTreeNode{Any}()
+RBTreeNode(d) = RBTreeNode{Any}(d)
+
+function Base.isequal(x::RBTreeNode{K}, y::RBTreeNode{K})where K
+   return (x.color == y.color && x.data == y.data)
+end
+
+Base.:(==)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = isequal(x, y)
+
+Base.:(!=)(x::RBTreeNode{K}, y::RBTreeNode{K}) where K = !isequal(x, y)
+
+function create_null_node(K::Type)
+    node = RBTreeNode{K}()
+    node.color = false
+    return node
+end 
+
+mutable struct RBTree{K}
+    root::RBTreeNode{K}
+    Nil::RBTreeNode{K}
+
+    function RBTree{K}() where K 
+        new{K}(create_null_node(K), create_null_node(K))
+    end
+end
+
+RBTree() = RBTree{Any}()
+
+function search_node(tree::RBTree{K}, d::K) where K
+    node = tree.root
+    while node != tree.Nil && d != node.data
```

Comment:
`node != tree.Nil`
is the same as:
```
(node.color === false && node.data === node.parent===node.left_child === node.right_child === nothing)
```
Potentially instead of storing a explict `Nil` in the tree to compare to we can just define a `is_nil` function that does that check.

Or conversly, we could always make sure the node is a reference to the tree's `Nil`,
and then do `!==` which would only do a pointer equality check.


if we did get rid of the `Nil` element, then we could also get rid of the `Tree` type and just have the `Node` type, since the tree would be described completely by its root.

Reply:
The implementation of Red-Black Tree becomes complicated when `T.nil` is not used, especially in cases related to `delete`. Let's keep this for the this now. If I find a good workaround, I'll put a PR later.

Reply:
Can we do this then:

> Or conversly, we could always make sure the node is a reference to the tree's Nil,
and then do !== which would only do a pointer equality check.